cs/src/core/Transcoder.cs

```diff
@@ -93,16 +110,76 @@ public Transcoder()
         /// <param name="writer">Writer instance</param>
         public void Transcode(R reader, W writer)
         {
-            transcode[0](reader, writer);
+            helper.Transcode(reader, writer);
         }
 
-        Action<R, W>[] Generate<S>(S schema)
+        TranscoderHelper helper;
+
+        private class TranscoderHelper
         {
-            var parser = ParserFactory<R>.Create(schema);
-            return SerializerGeneratorFactory<R, W>.Create(
-                    (r, w, i) => transcode[i](r, w), schema)
-                .Generate(parser)
-                .Select(lambda => lambda.Compile()).ToArray();
+            readonly Action<R, W>[] transcode;
+
+            public TranscoderHelper(RuntimeSchema schema)
+            {
+                transcode = Generate(schema);
+            }
+
+            public TranscoderHelper(Type type)
+            {
+                transcode = Generate(type);
+            }
+
+            virtual public void Transcode(R reader, W writer)
+            {
+                transcode[0](reader, writer);
+            }
+
+            Action<R, W>[] Generate<S>(S schema)
+            {
+                var parser = ParserFactory<R>.Create(schema);
+                return SerializerGeneratorFactory<R, W>.Create(
+                        (r, w, i) => transcode[i](r, w), schema)
+                    .Generate(parser)
+                    .Select(lambda => lambda.Compile()).ToArray();
+            }
+        }
+
+        private class TwoPassTranscoderHelper<FPW> : TranscoderHelper
+        {
+            readonly Action<R, FPW>[] firstPassTranscode;
+
+            public TwoPassTranscoderHelper(RuntimeSchema schema):
+                base(schema)
+            {
+                firstPassTranscode = GenerateFirstPass(schema);
+            }
+
+            public TwoPassTranscoderHelper(Type type):
+                base(type)
+            {
+                firstPassTranscode = GenerateFirstPass(type);
+            }
+
+            public override void Transcode(R reader, W writer)
+            {
+                var firstPassWriter = ((ITwoPassProtocolWriter)writer).GetFirstPassWriter();
+                if (firstPassWriter != null)
+                {
+                    R clonedReader = ((ICloneable<R>)reader).Clone();
```

Comment:
Not all readers are clonable. This should fail more gracefully when they aren't. We'll also want to check that the writer actually implemented `ITwoPassProtocolWriter` in the ctor like in `Serializer`.